src/widgets/index/index.ts

```diff
@@ -342,6 +344,105 @@ const index = (props: IndexProps): Index => {
         // it at the index level because it's either: all of them or none of them
         // that are stalled. The queries are performed into a single network request.
         instantSearchInstance.scheduleStalledRender();
+
+        if (__DEV__) {
+          type StateToWidgets = {
+            [TParameter in keyof IndexUiState]: Array<Widget['$$type']>;
+          };
+
+          const stateToWidgetsMap: StateToWidgets = {
+            query: ['ais.searchBox', 'ais.autocomplete', 'ais.voiceSearch'],
+            refinementList: ['ais.refinementList'],
+            menu: ['ais.menu'],
+            hierarchicalMenu: ['ais.hierarchicalMenu'],
+            numericMenu: ['ais.numericMenu'],
+            ratingMenu: ['ais.ratingMenu'],
+            range: ['ais.range'],
+            toggle: ['ais.toggleRefinement'],
+            geoSearch: ['ais.geoSearch'],
+            sortBy: ['ais.sortBy'],
+            page: ['ais.pagination', 'ais.infiniteHits'],
+            hitsPerPage: ['ais.hitsPerPage'],
+            configure: ['ais.configure'],
+          };
+
+          const mountedWidgets = this.getWidgets()
+            .map(widget => widget.$$type)
+            .filter(Boolean);
+
+          const missingWidgets = Object.keys(localUiState).reduce<
+            Array<Partial<StateToWidgets>>
+          >((acc, parameter) => {
+            const requiredWidgets: Array<Widget['$$type']> =
+              stateToWidgetsMap[parameter];
+
+            if (
+              !requiredWidgets.some(requiredWidget =>
+                mountedWidgets.includes(requiredWidget)
+              )
+            ) {
+              acc.push({
+                [parameter]: stateToWidgetsMap[parameter],
+              });
+            }
+
+            return acc;
+          }, []);
+
+          warning(
+            missingWidgets.length === 0,
+            `The UI state for the index "${this.getIndexId()}" is not consistent with the widgets mounted.
+
+This can happen when the UI state is specified via \`initialUiState\` or \`routing\` but that the widgets responsible for this state were not added. This results in those query parameters not being sent to the API.
+
+To fully reflect the state, some widgets need to be added to the index "${this.getIndexId()}":
+
+${missingWidgets
+  .map(widget => {
+    const stateParameter = Object.keys(widget)[0];
+    const neededWidgets = stateToWidgetsMap[stateParameter]
```

Comment:
We already have access to the needed widgets inside `widget` if I'm not mistaken.